### 1. Import Lib & Read file 

In [ ]:
import pandas as pd
import numpy as np

pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

RAW_DATA_PATH = r'D:\MinhDuckDesktop\PROJECTS\DA\VietNam_Bank_Churn_Analysis\data\raw\bank_churn_dataset_80k.csv'

df_raw = pd.read_csv(RAW_DATA_PATH)
df_raw.head()

,id,full_name,credit_sco,gender,age,occupation,balance,monthly_ir,address,origin_province,tenure_ye,married,nums_card,nums_service,active_member,last_active_date,last_transaction_month,created_date,exit,customer_segment,engagement_score,loyalty_level,digital_behavior,risk_score,risk_segment,cluster_group
0,1,Đặng Văn Vũ,725,male,55,Chủ Doanh nghiệp nhỏ,177306004,121000000,Phường An Hội (BT),TP. Hồ Chí Minh,0,2,4,8,True,03/04/2025,91038993,27/02/2025,False,Priority,90,Bronze,mobile,0.0359,Low,4
1,2,Bùi Hữu Phúc,689,male,45,Nội trợ/Sinh viên,1927416,5000000,Phường Bến Thành (Q1),Đồng Nai,3,1,3,2,True,14/05/2025,3255569,29/07/2021,False,Mass,63,Gold,mobile,0.2664,Low,2
2,3,Lê Văn Mai,702,female,44,Chủ Doanh nghiệp nhỏ,304931745,109000000,Phường Hòa Bình (Q11),TP. Hồ Chí Minh,4,0,2,8,False,10/09/2021,0,03/03/2021,False,Priority,36,Silver,offline,0.1343,Low,4
3,4,Dương Trần Nhiên,766,male,44,Chủ Doanh nghiệp nhỏ,50615501,79000000,Phường 7 (Q7),TP. Hồ Chí Minh,3,0,4,3,False,04/03/2025,0,12/02/2022,False,Priority,23,Bronze,offline,0.2185,Low,4
4,5,Dương Thu Linh,677,female,77,Giáo viên/Giảng viên,40532432,25000000,Phường Thắng Lợi (Q10),TP. Hồ Chí Minh,2,1,4,3,False,06/08/2022,0,24/07/2022,False,Emerging,23,Bronze,offline,0.2942,Low,2


### 2. Check size & datatype

In [4]:
df_raw.shape
df_raw.info(verbose=True, show_counts=True)

<class 'pandas.DataFrame'>
RangeIndex: 80000 entries, 0 to 79999
Data columns (total 26 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   id                      80000 non-null  int64  
 1   full_name               80000 non-null  str    
 2   credit_sco              80000 non-null  int64  
 3   gender                  80000 non-null  str    
 4   age                     80000 non-null  int64  
 5   occupation              80000 non-null  str    
 6   balance                 80000 non-null  int64  
 7   monthly_ir              80000 non-null  int64  
 8   address                 80000 non-null  str    
 9   origin_province         80000 non-null  str    
 10  tenure_ye               80000 non-null  int64  
 11  married                 80000 non-null  int64  
 12  nums_card               80000 non-null  int64  
 13  nums_service            80000 non-null  int64  
 14  active_member           80000 non-null  bool   
 

### 3. Missing / Duplicates & Describe

In [5]:
print("Missing values per column:\n", df_raw.isnull().sum())
print("\nDuplicate rows count:", df_raw.duplicated().sum())
print("\nDescriptive statistics:")
df_raw.describe().round(2)

Missing values per column:
 id                        0
full_name                 0
credit_sco                0
gender                    0
age                       0
occupation                0
balance                   0
monthly_ir                0
address                   0
origin_province           0
tenure_ye                 0
married                   0
nums_card                 0
nums_service              0
active_member             0
last_active_date          0
last_transaction_month    0
created_date              0
exit                      0
customer_segment          0
engagement_score          0
loyalty_level             0
digital_behavior          0
risk_score                0
risk_segment              0
cluster_group             0
dtype: int64

Duplicate rows count: 0

Descriptive statistics:


,id,credit_sco,age,balance,monthly_ir,tenure_ye,married,nums_card,nums_service,last_transaction_month,engagement_score,risk_score,cluster_group
count,80000.00,80000.00,80000.00,8.000000e+04,80000.0,80000.00,80000.00,80000.00,80000.00,8.000000e+04,80000.00,80000.00,80000.00
mean,40000.50,684.32,49.38,6.118320e+07,37788200.0,1.78,0.97,2.63,3.56,8.831755e+06,31.73,0.27,1.97
std,23094.16,51.17,15.07,8.868620e+07,36551066.2,1.32,0.69,1.22,2.00,3.378683e+07,23.47,0.09,1.04
min,1.00,495.00,20.00,4.086630e+05,3000000.0,0.00,0.00,1.00,1.00,0.000000e+00,7.00,0.02,1.00
25%,20000.75,649.00,38.00,1.462592e+07,14000000.0,1.00,1.00,2.00,2.00,0.000000e+00,16.00,0.20,1.00
50%,40000.50,682.00,47.00,3.090276e+07,23000000.0,2.00,1.00,2.00,3.00,0.000000e+00,22.00,0.28,2.00
75%,60000.25,718.00,58.00,6.960994e+07,47000000.0,3.00,1.00,3.00,5.00,0.000000e+00,38.00,0.34,2.00
max,80000.00,800.00,90.00,1.000000e+09,216000000.0,4.00,3.00,5.00,8.00,6.000000e+08,100.00,0.54,4.00


### 4. Check categorical data

In [6]:
cat_cols = df_raw.select_dtypes(include=['object', 'category', 'string']).columns

for col in cat_cols:
    print(f"Column: {col}")
    print(df_raw[col].value_counts())
    print("\n")

Column: full_name
full_name
Lê Minh Quang         42
Phạm Anh Quang        41
Vũ Đức Quang          41
Mai Thanh Quang       40
Mai Hữu Quang         38
Bùi Anh Quang         38
Lê Thu Quang          37
Trần Minh Quang       37
Bùi Minh Quang        37
Dương Hoàng Quang     37
Trần Thị Quang        36
Mai Đức Quang         36
Vũ Thu Quang          35
Dương Văn Quang       35
Phạm Thanh Quang      34
Lê Kim Quang          34
Đặng Thanh Lan        33
Dương Văn Vân         33
Đặng Quang Quang      33
Trần Gia Quang        33
Bùi Xuân Quang        33
Vũ Thanh Quang        33
Vũ Anh Quang          32
Bùi Thị Quang         32
Vũ Minh Ánh           32
Bùi Hữu Quang         32
Trần Hoàng Quang      32
Đặng Thu Quang        32
Trần Kim Quang        32
Nguyễn Thanh Quang    32
Bùi Hoàng Quang       32
Phạm Minh Quang       32
Lê Trần Quang         31
Nguyễn Thu Quang      31
Lê Thanh Quang        31
Hoàng Thị Quang       31
Hoàng Trần Quang      31
Hoàng Quang Quang     31
Dương Trần Quang      

### 5. Data Quality Assessment Report

**Kích thước & Cấu trúc (Volume & Schema):**

- Tập dữ liệu gồm 80,000 dòng và 26 cột.

- Đầy đủ các trường thuộc tính từ thông tin cá nhân, chỉ số tài chính, lịch sử giao dịch cho đến phân nhóm rủi ro (risk_segment, cluster_group).

**Độ toàn vẹn (Integrity & Health Check):**

- Missing values: 0 (Toàn bộ 80,000 dòng ở 26 cột đều non-null).

**Phân phối chỉ số số (Numeric summary):**

- Độ tuổi (age): Từ 20 đến 90 tuổi (trung bình 49.38), hợp lý.

- Điểm tín dụng (credit_score): Từ 495 đến 800.

- Số dư (balance): Không có giá trị âm.

### 6. Export raw data to DataLake

In [12]:
import os
import dotenv
import sys
sys.path.append(os.path.abspath(".."))
from src.db_connection import get_db_engine 

dotenv.load_dotenv()

table_name = os.getenv('TB_NAME_RAW_BANK_CHURN')
DB_NAME_DATALAKE = os.getenv('DB_NAME_DATALAKE')
engine = get_db_engine(DB_NAME_DATALAKE)
try:
    df_raw.to_sql(name=table_name, con=engine, if_exists='replace', index=False)
except Exception as e:
    print(f"An error occurred: {e}")